spec/lib/date_name_spec.rb

```diff
@@ -0,0 +1,103 @@
+# encoding: utf-8
+require File.expand_path('../spec_helper', File.dirname(__FILE__))
+require 'core_ext'
+
+
+describe SmartName do
+
+  describe '#key' do
+    it 'should remove spaces and non-key chars' do
+      n=' January -_& 1, 1999'.to_name
+      n.pretty_key.should == 'Date: 01/01/99'
+      n.key.should == 'DACVm7A'
+    end
+
+    it 'should do Y M D' do
+      n='1999-1-31'.to_name
+      n.pretty_key.should == 'Date: 01/31/99'
+      n.key.should == 'DACVnCg'
+    end
+
+    it 'should not do Y D M' do
+      n='1999-31-1'.to_name
+      n.pretty_key.should == 'Long: 1999::Long: 31::Long: 1'
+      n.key.should == 'LAAAHzw_LAAAAHw_LAAAAAQ'
+    end
+
+    it 'should do M D Y based on range' do
+      n='1 31 1999'.to_name
+      n.pretty_key.should == 'Date: 01/31/99'
+      n.key.should == 'DACVnCg'
+    end
+
+    it 'should default to D M Y' do
+      n='1 10 1999'.to_name
+      n.pretty_key.should == 'Date: 10/01/99'
+      n.key.should == 'DACVn/Q'
+    end
+
+    it 'should ignore week day' do
+      n='Friday, Jan 10, 1999'.to_name
+      n.pretty_key.should == 'Date: 10/01/99'
+      n.key.should == 'DACVn/Q'
+    end
+
+    it 'should not do year in the middle' do
+      n='Jan 1999 10'.to_name
+      n.pretty_key.should == 'jan::Long: 1999::Long: 10'
+      n.key.should == 'jan_LAAAHzw_LAAAACg'
+    end
+
+    it 'should not match unless right after ignored week day' do
+      n='Friday, on Jan 10, 1999'.to_name
+      n.pretty_key.should == 'friday::on::jan::Long: 10::Long: 1999'
+      n.key.should == 'friday_on_jan_LAAAACg_LAAAHzw'
+    end
+
+    it 'matches after other keys' do
+      n='some, key Jan 10, 1999'.to_name
+      n.pretty_key.should == 'some::key::Date: 10/01/99'
+      n.key.should == 'some_key_DACVn/Q'
+    end
+
+    it 'matches after number keys' do
+      n='10 20 30, Jan 10, 1999'.to_name
+      n.pretty_key.should == 'Long: 10::Long: 20::Long: 30::Date: 10/01/99'
+      n.key.should == 'LAAAACg_LAAAAFA_LAAAAHg_DACVn/Q'
+    end
+
+    it 'matches after number keys, including time' do
+      n='10 20 30, Jan 10, 1999 10am'.to_name
+      n.pretty_key.should == 'Long: 10::Long: 20::Long: 30::Datetime: Fri Oct  1 10:00:00 1999'
+      #n.key.should == 'LAAAACg_LAAAAFA_LAAAAHg_TAAAAADf0zHA' # Base64 encoding is different only in point version at semaphore
+    end
+
+    it 'matches after number keys, including time' do
+      n='10 20 30, Jan 10, 1999 10AM'.to_name
+      n.pretty_key.should == 'Long: 10::Long: 20::Long: 30::Datetime: Fri Oct  1 10:00:00 1999'
+      #n.key.should == 'LAAAACg_LAAAAFA_LAAAAHg_TAAAAADf0zHA' # Base64 encoding is different only in point version at semaphore
+    end
+
+    it 'matches include 24 hour time' do
+      n='Jan 10, 1999 22:00:00'.to_name
+      n.pretty_key.should == 'Datetime: Fri Oct  1 22:00:00 1999'
+      #n.key.should == 'TAAAAADf1dTA' # Base64 encoding is different only in point version at semaphore
```

Comment:
I suspect the problem here could be related to time zone handling.  The result may in fact depend on your time zone setting and the difference may be that offset.  I suppose if you get different results from both me and semaphore that means the main issue is timezone.  The tests could factor that out, but it isn't clear what we really want here.  Clearly we don't want keys to be different depending on your timezone since they need to be stable for all users who will not always have the same default timezone.  We probably get the server timezone as it is, but the probably should be UTC based.